### PR TITLE
Fix to scale ages based on supplied span_scale

### PR DESCRIPTION
**Pull request by drew-sinha**
Previously, supplying an age_scale to worm_data.read_worms wouldn't scale any feature labeled by age (age, ghost_age, adult_age) per worm_data.read_summary_data:576. This fix allows ages to be scaled by the supplied scaling factor to read_worms/read_summary_data as well.

### CODE REVIEW

**Comment by zpincus** [comment]
Wait, what error does this fix? Summary data (i.e. per-worm measurements like "lifespan", stored as `worm.lifespan`) shouldn't have any "age" or "ghost_age" or similar columns. Such columns would be per-timepoint data, not per-worm (i.e. stored as `worm.td.age`: a worm has one lifespan but has a specific age at each timepoint). Per-timepoint data is read in by `load_worms`, which has an `age_scale` parameter already. 

Note also that `load_worms` allows one to pass in summary data file(s) to also load, and will use the supplied `age_scale` parameter as the `span_scale` for loading in the summary data. So that's one-stop shopping there.

**Comment by drew-sinha** [comment]
Ok - wow that was dumb. I figured out what the original issue was.....

I called read_worms with a timecourse file + age_scale (1/24 for days) but no summary glob/file. I expected all my ages/spans to be in days, but instead were in hours. I localized the problem to the call to Worm.calculate_ages_and_spans (at 108) where the ages/spans that are calculated are in hours in this function; after the call, the age_scale never gets applied. The updated fix applies the age_scale outside the calculate_ages_and_spans call (to keep the logic of the function consistent).

**Comment by drew-sinha** [comment]
One more fix for my bug - put the scaled span data under the worm and ages under the w.td.

File: elegant/worm_data.py [hunk 1]
```diff
@@ -106,6 +106,16 @@ def get_genotype(path):
             for w in worms:
                 try:
                     w.calculate_ages_and_spans()
+                    for feature_name in dir(w.td):
```
Comment (zpincus) [inline elegant/worm_data.py:109]:
```
>>> t = worm_data._TimecourseData()
>>> dir(t)
['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_items',
 '_keys']
```
use `w.td._keys()` or `w.td._items()` instead:
```
for feature_name, feature_vals in w.td._items():
    ...
```

File: elegant/worm_data.py [hunk 2]
```diff
@@ -106,6 +106,16 @@ def get_genotype(path):
...
+                        if feature_name == 'age' or feature_name.endswith('_age'):
+                            feature_vals = getattr(w.td, feature_name)
+                            feature_vals *= age_scale
```
Comment (zpincus) [inline elegant/worm_data.py:112]: If you look at how the `Worm` constructor does this, it explicitly casts the values to a numpy array before trying to multiply -- what if a feature happens to be a list for some reason?

**Review by zpincus — COMMENTED** [review]
Aside from the code nitpicks below, the biggest issue is a logical one: this will scale *twice* any existing `age` or `span` data that came in from the files and which don't get overwritten by `calculate_ages_and_spans()`. The right solution is to have that function apply the correct scaling to the new ages/spans it calculates. But even that is tricky, because we don't know the desired output units: just an undefined scale factor.

Probably a thorough rethink is required. I'm leaning toward removing the `scale` parameters and replacing them with a pair of `input_time_units` and `desired_time_unit`or something. Or just have everything always in hours by default, and then have a single `rescale_time` function?

File: elegant/worm_data.py (hunk 2 above)
Comment (drew-sinha) [reply elegant/worm_data.py:112]: Re the code comment: yeah, explicit casting would probably be the best....

As far as the larger logical issue, I'm partial to the latter option. It seems simpler to not worry about the scaling nonsense until the last possible minute (which I think could be reasonably handled with a rescale_time function). I'm guessing the constructor would spin up, calculate the ages/spans, and then rescale once at the end. Or would the user have to call it explicitly?

**Comment by zpincus** [comment]
Fixed (I hope) in 34cd756 and 7342beb.

**Pull request closed without merging**